### PR TITLE
Deprecating subclassing of Plugin

### PR DESCRIPTION
**Pull request by jglick**
[Discussion](https://groups.google.com/forum/#!msg/jenkinsci-dev/V394esGVs88/WvSN9EGdBwAJ). Have I missed any use cases which are not handled just as well or better by newer APIs?

@reviewbybees

### CODE REVIEW

**Comment by olivergondza** [comment]
+1

**Comment by amuniz** [comment]
Should not be the class itself deprecated also (in order to avoid `extends` of it)?

**Comment by KostyaSha** [comment]
@amuniz +1 also expected deprecation on class level. @jglick why you decided introduce constructor?

File: core/src/main/java/hudson/Plugin.java [hunk 1]
```diff
@@ -79,6 +82,22 @@
...
+     * <li>{@link #start} and {@link #postInitialize} can be replaced by {@link Initializer} (or {@link ItemListener#onLoaded})
+     * <li>{@link #stop} can be replaced by {@link Terminator}
+     * <li>{@link #setServletContext} can be replaced by {@link Jenkins#servletContext}
+     * </ul>
+     * Note that every plugin gets a {@link DummyImpl} by default,
+     * which will still route the URL space, serve {@link #getWrapper}, and so on.
+     * @deprecated Use more modern APIs rather than subclassing.
+     */
+    @Deprecated
+    protected Plugin() {}
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/Plugin.java]: :bug:  May cause regressions in the existing plugin code, which utilizes public constructors. Not sure the case is realistic.

I would vote for adding a public constructor and marking it as Restricted

**Comment by jglick** [comment]
> Should not be the class itself deprecated also (in order to avoid `extends` of it)?

No, deprecating the constructor suffices, since any subclass must call the `super()` constructor, triggering a deprecation warning when compiling against the new core. And I did *not* want non-subclass usages of the class to be deprecated, as there are valid (if uncommon) use cases for calling methods on it, or simply checking whether `getPlugin("something") != null`.

> May cause regressions in the existing plugin code, which utilizes public constructors.

Not possible. The class was already `abstract`. Therefore its (implicit) `public` constructor could only have been called by subclasses, which can still do so.

**Comment by oleg-nenashev** [comment]
> Not possible. The class was already abstract. Therefore its (implicit) public constructor could only have been called by subclasses, which can still do so.

Agreed, Have not noticed "abstract".

:bee:

**Pull request merged**